drivers/bluetooth/hci/userchan.c

```diff
@@ -155,6 +155,11 @@ static uint16_t packet_len(const uint8_t *buf)
 		return 0;
 	}
 
+	/* Prevent overflow (e.g. when we've received a partial header so payload_len is junk) */
+	if (UINT16_MAX - sizeof(type) - header_len < payload_len) {
```

Comment:
Would it not be better to actually verify if we recieved a partial header? Taking as reference Bluetooth Spec v5.4 Vol 4 Part E you will find that the payload length can be either have a length of 1 byte (H4_CMD, H4_SCO, H4_EVT) or 2 bytes (H4_ACL, H4_ISO) so I am not sure if this work in two edge cases:

   1. When you recieve a fragmented H4 frame that has a payload of maximum 1 byte (i.e., UINT8_MAX)
   2. When you recieve a fragmented H4 frame with just the first payload byte (i.e. H4_ACL, H4_ISO) with a second invalid payload byte that. E.g., Payload len byte 1 = 0x01, Payload len byte 2 (invalid due to fragmentation) = 0xAB. This would bypass your oveflow check and assume you have a payload of size 0x01AB.
   
You could instead use the current size of your frame, and the frame type to verify if the frame payload length is not complete before you actually decode the H4 packet len with `packet_len`.

Reply:
Example suggestion
```c
static int packet_len(const uint8_t *buf, const uint8_t buf_size, uint16_t * packet_len)
{
	uint16_t payload_len = 0;
	uint8_t header_len = 0;
	const uint8_t type = buf[0];
	const uint8_t *hdr = &buf[sizeof(type)];
	
	if (buf_size ==  sizeof(type)) {
            return -EINVAL;
	}

	switch (type) {
	  case H4_CMD: {
		  const struct bt_hci_cmd_hdr *cmd = (const struct bt_hci_cmd_hdr *)hdr;
  
		  /* Parameter Total Length */
		  payload_len = cmd->param_len;
		  header_len = BT_HCI_CMD_HDR_SIZE;
		  
		  break;
	  }
	    	// rest of the switch cases
	  }

	  if (buf_size  < header_len ) {
	      return -EINVAL
	  }
	  
	  *packet_len = sizeof(type) + header_len + payload_len;
}
```

Reply:
Yea I initially did something along those lines, but then thought this way is less code. The way I've done it the length is sometimes "wrong" but is always greater than or equal to the true value. I check it never overflows, and if it has a full header it's correct, otherwise it's the true header length plus some random positive value, so it receives more until it does have a full header. The case where it sends just the initial type byte then the rest of the packet is in fact exactly the situation that motivated this patch.

Happy to change to the above if preferred though.

Reply:
Something along the lines of what @vChavezB suggested sounds like a better approach to me. That said, we have several examples of H4 decoding in the tree, so might make sense to just follow the same (proven) logic as used there:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/bluetooth/hci/h4.c
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/bluetooth/hci_uart/src/main.c

Reply:
I have changed it to roughly what @vChavezB suggests. I'm not sure it's worth refactoring both files to share with those other bits, because the relevant code is not very complicated and in this specific usage has the unusual requirement to not go wrong if the data is incomplete.